Student Performance Indicator (SPI)


 Life cycle of machine learning 
- understanding the problem statement 
- data collection
- data checks to perform 
- exploratory data analysis (EDA)
- Data pre-processing 
- model training 
- choose best model


1) Problem Statement
- This project explores the impact of various factors like gender, ethnicity, parents' education level, lunch, and test preparation courses on students' test scores.

2) For data collection
- the project uses a dataset from Kaggle, available at this link: https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977. 
- This dataset has 8 columns and 1000 rows.

2.1) Import Data and Required Packages
- Importing Pandas, Numpy, Matplotlib, Seaborn and Warings Library.

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

Import the CSV Data as Pandas DataFrame

In [2]:
df = pd.read_csv('data/stud.csv')

to show top five records

In [4]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

 to show the shape of the data

In [11]:
df.shape

(1000, 8)

2.2) Dataset information
- gender : sex of students -> (Male/female)
- race/ethnicity : ethnicity of students -> (Group A, B,C, D,E)
- parental level of education : parents' final education ->(bachelor's degree,some college,master's degree,associate's degree,high school)
- lunch : having lunch before test (standard or free/reduced)
- test preparation course : complete or not complete before test
- math score
- reading score
- writing score
3. Data Checks to perform
- Check Missing values
- Check Duplicates
- Check data type
- Check the number of unique values of each column
- Check statistics of data set
- Check various categories present in the different categorical column

In [9]:
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

3.1) Check missing values

- there are no missing values in the data set

3.2) Check Duplicate values

In [12]:
df.duplicated().sum()


0

there is no any duplicate values

3.2) Check the data types

In [13]:
# Check Null and Dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


3.4 Checking the number of unique values of each column


In [14]:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

3.5 Check statistics of data set

In [15]:
df.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

INSIGHT form above
- Close Average Scores: The average scores for math, reading, and writing are closely aligned, ranging between 66 and 68.05, indicating a general consistency in performance levels across these subjects.


- Similar Variability in Scores: The standard deviations for all three subjects are in a narrow range between 14.6 and 15.19, suggesting that the scores in math, reading, and writing vary to a similar extent among students.


- Differences in Minimum Scores: There is a notable difference in the minimum scores across subjects: math has the lowest minimum score of 0, while writing has a higher minimum of 10, and reading has an even higher minimum score of 17. This indicates that the lowest performance in math is significantly below that in reading and writing.

In [ ]:
3.7) Expl